scipy/_lib/_array_api.py

```diff
@@ -195,6 +195,89 @@ def _asarray(
     return array
 
 
+def _resolve_device(array, xp_out, device=None):
+    if device is not None:
+        return device
+
+    xp_in = array_namespace(array)
+
+    # Here is what the logic is intended to be
+    # if is_numpy(xp_out) or is_array_api_strict(xp_out) or is_cupy(xp_out):
+    #     # These libraries only support one type of device
+    #     return xp_out.__array_namespace_info__().default_device()
+    # else:
+    #     return xp_in.device(array)
+
+    # but there are many issues with device support right now. For example:
+    # - NumPy `from_dlpack` only understands device='cpu' (not a device object)
+    # - CuPy `from_dlpack` doesn't work
+    # - the format of the return of `device` seems different in different libraries
+
+    # So in the meantime:
+    if is_numpy(xp_out):
+        return "cpu"
+    elif is_array_api_strict(xp_out):
+        return xp_out.__array_namespace_info__().default_device()
+    else:
+        return None
+
+
+def xp_asarray(
+        array: ArrayLike,
+        dtype: Any = None,
+        order: Literal['K', 'A', 'C', 'F'] | None = None,
+        copy: bool | None = None,
+        *,
+        xp: ModuleType | None = None,
+        check_finite: bool = False,
+        subok: bool = False,
+    ) -> Array:
+    """SciPy-specific replacement for `xp.asarray` with `order`, `check_finite`,
+    `subok`, and automatic array type conversion and device transfer.
```

Comment:
When I wrote that, I was trying to make device transfers work as an experiment. Regardless of whether SciPy actually does this or not, I thought it would be handy to make this function do whatever we might ask of it. 

I know in https://github.com/scipy/scipy/issues/18286#issue-1664855687, you declared that we wouldn't transfer.
> For as-yet-unsupported GPU execution when hitting compiled code, we will raise exceptions. The alternative considered was to transfer to CPU, execute, and transfer back (e.g., for PyTorch). A pro of doing that would be that everything works, and there may still be performance gains. A con is that it silently does device transfers, usually not a good idea.

At some point it would be nice to discuss this. I'd like to understand why it's "not a good idea" if it's faster overall and just works. I haven't brought it up explicitly because `from_dlpack` with device transfers is still not well supported, so even if we do want to allow SciPy to do the transfers, it won't work.

But for now, that comment can certainly be removed.